### ISSUE TITLE
confirm output reliability across platforms

### ISSUE THREAD

**Issue by alanngnet** [alanngnet/CoverHunterMPS#29]
We have observed significantly inconsistent training results between specific computers. Investigate and resolve the issue(s). Could be computer-specific problems, or Pytorch bugs, rather than project-specific problems?

**Comment by alanngnet** [comment]
Remaining task is to verify fix on CUDA.

**Comment by alanngnet** [comment]
Latest discovery: covers80 trains deterministically after the latest patches, but another training dataset (our irish_reels_test data set) does not.

**Comment by alanngnet** [comment]
Useful discovery: Setting hyperparameter mode="defined" causes consistent deterministic training results. This may be sufficient to get back to work using Tunography (Irish trad) data because of the hiqh quality clipping of that data, so that for that data srouce, always starting chunks at the 0 second position is a good way to chunk. Whereas data sources with arbitrary delays before relevant music begins probably need mode="random" to be more reliable.

**Comment by alanngnet** [comment]
Verified still deterministic when using mode="defined" even with different chunk_frames and chunk_s settings.

**Comment by alanngnet** [comment]
Nope, still non-deterministic training in some situations. Not clear what factors define those situations.

**Comment by alanngnet** [comment]
Measured range of variability in training runs given same data and hyperparameters and number of epochs, but different random seeds. Across 4 runs, observed standard deviations of the 4th-highest mAP scores are around 5-7%, which is very significant given that fine-tuning work in the field of CSI is often aiming at few-percent improvements, especially since fine-tuning work needs to happen with smaller datasets that naturally have higher variability of training outcomes.

**Comment by alanngnet** [comment]
Progress report:

MPS runs are mostly deterministic using current code, even with `mode`: "random", but not all the time, still unclear what leads to non-determinism on MPS.

Now that @samuel-gauthier is back at work on this issue: On CUDA the model is *never* deterministic. Current evidence suggests something in the optimizer stage is causing CUDA-specific non-determinism.

@samuel-gauthier has confirmed that the math precision-reduction changes I made in the very first CoverHunterMPS commits are apparently *not* relevant to the problem and also do not hurt overall training results:
Commit abe657d
src/loss.py: Old lines 87, 106
src/model.py:New line 232

Commit af0d21a 
src/loss.py: New line 122

**Comment by alanngnet** [comment]
Demoting this ticket from critical P0 to important P1 now that train_tune script mitigates non-deterministic behavior by averaging across multiple seeds. Inconsistent non-deterministic behavior still observed on MPS, but I noted a possible pattern of deterministic results when there's a system reboot just before both runs.
